src/System.Private.CoreLib/src/System/Threading/Tasks/TaskFactory.cs

```diff
@@ -2308,8 +2317,12 @@ public CompleteOnInvokePromise(IList<Task> tasks) : base()
 
             public void Invoke(Task completingTask)
             {
-                if (m_firstTaskAlreadyCompleted == 0 &&
-                    Interlocked.Exchange(ref m_firstTaskAlreadyCompleted, 1) == 0)
+                int isSyncBlockingFlag = IsSyncBlockingFlag;
+                // Mask off the sync-blocking flag
+                int firstTaskAlreadyCompleted = m_firstTaskAlreadyCompleted & 0b_01;
+
+                if (firstTaskAlreadyCompleted == 0 &&
+                    Interlocked.Exchange(ref m_firstTaskAlreadyCompleted, isSyncBlockingFlag & 0b_01) == isSyncBlockingFlag)
```

Comment:
I'm not understanding this. isSyncBlockingFlag is "m_firstTaskAlreadyCompleted & 0b_10", so this is "m_firstTaskAlreadyCompleted & 0b_10 & 0b01"... doesn't that just make it 0?

Reply:
Fixed